File: tools/release/package_cli.py
#!/usr/bin/env python3
"""Create and smoke-test one deterministic Caudex CLI release archive."""

import argparse
import hashlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import tarfile
import tempfile
import zipfile
import gzip
import io


VERSION = "0.1.0"
FIXED_MTIME = 1_767_225_600  # 2026-01-01T00:00:00Z


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as source:
        for chunk in iter(lambda: source.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def run(command, cwd: Path, env=None) -> str:
    result = subprocess.run(command, cwd=cwd, env=env, text=True, capture_output=True)
    if result.returncode:
        raise SystemExit(
            f"command failed ({result.returncode}): {' '.join(map(str, command))}\n"
            f"stdout: {result.stdout}\nstderr: {result.stderr}"
        )
    return result.stdout


def add_tar_entry(archive: tarfile.TarFile, path: Path, relative: str) -> None:
    data = path.read_bytes()
    info = tarfile.TarInfo(relative)
    info.size = len(data)
    info.mtime = FIXED_MTIME
    info.uid = 0
    info.gid = 0
    info.uname = ""
    info.gname = ""
    info.mode = 0o755 if relative.endswith("/caudex") else 0o644
    archive.addfile(info, io.BytesIO(data))


def write_tar_gz(stage: Path, output: Path, stem: str) -> None:
    output.parent.mkdir(parents=True, exist_ok=True)
    with output.open("wb") as raw:
        with gzip.GzipFile(fileobj=raw, mode="wb", mtime=0, filename="") as compressed:
            with tarfile.open(fileobj=compressed, mode="w", format=tarfile.PAX_FORMAT) as archive:
                for path in sorted(stage.rglob("*")):
                    if path.is_file():
                        relative = f"{stem}/{str(path.relative_to(stage)).replace(os.sep, '/')}"
                        add_tar_entry(archive, path, relative)


def write_zip(stage: Path, output: Path, stem: str) -> None:
    output.parent.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(output, "w", compression=zipfile.ZIP_DEFLATED, compresslevel=9) as archive:
        for path in sorted(stage.rglob("*")):
            if path.is_file():
                relative = str(path.relative_to(stage)).replace(os.sep, "/")
                info = zipfile.ZipInfo(relative, date_time=(2026, 1, 1, 0, 0, 0))
                info.filename = f"{stem}/{relative}"
                info.create_system = 3
                info.external_attr = (0o755 if relative.endswith("/caudex.exe") else 0o644) << 16
                archive.writestr(info, path.read_bytes())


def extract(archive: Path, destination: Path, is_windows: bool) -> Path:
    destination.mkdir(parents=True, exist_ok=True)
    if is_windows:
        with zipfile.ZipFile(archive) as source:
            source.extractall(destination)
    else:
        with tarfile.open(archive, "r:gz") as source:
            source.extractall(destination)
    roots = [path for path in destination.iterdir() if path.is_dir()]
    if len(roots) != 1:
        raise SystemExit(f"archive must contain one top-level directory: {roots}")
    return roots[0]


def smoke(root: Path, target: str) -> None:
    executable = root / ("caudex.exe" if target == "x86_64-windows-gnu" else "caudex")
    if not executable.is_file():
        raise SystemExit(f"missing executable: {executable}")
    if target == "x86_64-windows-gnu" and os.name != "nt":
        # Makes fixture tests on Unix hosts exercise archive contents too.
        executable.chmod(0o755)
    if target == "x86_64-windows-gnu" and not (root / "sqlite3.dll").is_file():
        raise SystemExit("Windows archive is missing sqlite3.dll")

    environment = os.environ.copy()
    if target == "x86_64-windows-gnu":
        # The executable directory is the only user-controlled DLL search path.
        environment["PATH"] = str(root)
    version_output = run([str(executable), "version"], root, environment)
    if not version_output.startswith("caudex 0.1.0\n"):
        raise SystemExit(f"unexpected CLI version output: {version_output}")
    help_text = run([str(executable), "--help"], root, environment)
    if "database backup" not in help_text or "history last" not in help_text:
        raise SystemExit("extracted executable returned incomplete help")

    database = root / "smoke.sqlite"
    check = run([str(executable), "--database", str(database), "--format", "json", "database", "check"], root, environment)
    if '"integrity":"ok"' not in check:
        raise SystemExit(f"SQLite smoke check failed: {check}")
    started = run(
        [
            str(executable), "--database", str(database), "--format", "json",
            "workout", "start", "--command-id", "release-smoke-start",
            "--workout", "release-smoke-workout", "--started-at", "2026-01-01T00:00:00Z",
            "--occurred-at", "2026-01-01T00:00:00Z",
        ],
        root,
        environment,
    )
    if '"workoutId":"release-smoke-workout"' not in started:
        raise SystemExit(f"workout smoke start failed: {started}")
    shown = run(
        [str(executable), "--database", str(database), "--format", "json", "workout", "show", "--workout", "release-smoke-workout"],
        root,
        environment,
    )
    if '"workoutId":"release-smoke-workout"' not in shown:
        raise SystemExit(f"workout smoke read failed: {shown}")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--target", required=True)
    parser.add_argument("--archive", choices=("tar.gz", "zip"), required=True)
    parser.add_argument("--binary", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--sqlite-version", required=True)
    parser.add_argument("--sqlite-linkage", required=True)
    parser.add_argument("--sqlite-source", required=True)
    parser.add_argument("--sqlite-library", type=Path)
    parser.add_argument("--sqlite-dll", type=Path)
    parser.add_argument("--sqlite-checksum")
    args = parser.parse_args()

    stem = args.output.name.removesuffix(".tar.gz").removesuffix(".zip")
    with tempfile.TemporaryDirectory(prefix="caudex-cli-package-") as temporary:
        stage_root = Path(temporary) / stem
        stage_root.mkdir()
        executable_name = "caudex.exe" if args.target == "x86_64-windows-gnu" else "caudex"
        shutil.copyfile(args.binary, stage_root / executable_name)
        shutil.copyfile("LICENSE", stage_root / "LICENSE")
        shutil.copyfile("NOTICE", stage_root / "NOTICE")
        shutil.copyfile("apps/caudex-cli/README.md", stage_root / "README.md")
        if args.sqlite_dll:
            shutil.copyfile(args.sqlite_dll, stage_root / "sqlite3.dll")

        sqlite_file = args.sqlite_library or args.sqlite_dll
        if sqlite_file is None or not sqlite_file.is_file():
            raise SystemExit("a SQLite library or DLL is required to record a checksum")
        metadata = {
            "schemaVersion": 1,
            "version": VERSION,
            "target": args.target,
            "optimize": "ReleaseSafe",
            "archiveStem": stem,
            "platformCodeSigned": False,
            "sqlite": {
                "version": args.sqlite_version,
                "linkage": args.sqlite_linkage,
                "source": args.sqlite_source,
                "checksum": {"algorithm": "sha256", "value": args.sqlite_checksum or sha256(sqlite_file)},
            },
        }
        (stage_root / "build-metadata.json").write_text(json.dumps(metadata, sort_keys=True, indent=2) + "\n")

        if args.archive == "zip":
            write_zip(stage_root, args.output, stem)
        else:
            write_tar_gz(stage_root, args.output, stem)

        with tempfile.TemporaryDirectory(prefix="caudex-cli-extract-") as extracted:
            extracted_root = extract(args.output, Path(extracted), args.archive == "zip")
            if extracted_root.name != stem:
                raise SystemExit(f"unexpected archive root: {extracted_root.name}")
            smoke(extracted_root, args.target)

    print(f"packaged and smoke-tested {args.output}")


if __name__ == "__main__":
    main()

File: tools/release/package_cli.mjs
#!/usr/bin/env node
import fs from "node:fs/promises";
import { chmodSync, existsSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { archiveEntries, sha256, writeTarGz, writeZip } from "./archive-utils.mjs";

const args = parseArgs(process.argv.slice(2));
const required = ["version", "target", "archive", "binary", "output", "sqlite-version", "sqlite-linkage", "sqlite-source"];
for (const name of required) if (!args[name]) throw new Error(`missing required argument --${name}`);
if (!["tar.gz", "zip"].includes(args.archive)) throw new Error(`unsupported archive format: ${args.archive}`);

const repository = path.resolve(import.meta.dirname, "../..");
const binary = path.resolve(args.binary);
const output = path.resolve(args.output);
const stem = path.basename(output).replace(/\.tar\.gz$|\.zip$/, "");
const windows = args.target.endsWith("windows-gnu");
const executableName = windows ? "caudex.exe" : "caudex";
const sqliteFile = args["sqlite-library"] ?? args["sqlite-dll"];
if (!sqliteFile || !existsSync(path.resolve(sqliteFile))) throw new Error("a SQLite library or DLL is required to record a checksum");

const temporary = await fs.mkdtemp(path.join(os.tmpdir(), "caudex-cli-package-"));
try {
  const stage = path.join(temporary, stem);
  await fs.mkdir(stage, { recursive: true });
  await fs.copyFile(binary, path.join(stage, executableName));
  await fs.copyFile(path.join(repository, "LICENSE"), path.join(stage, "LICENSE"));
  await fs.copyFile(path.join(repository, "NOTICE"), path.join(stage, "NOTICE"));
  await fs.copyFile(path.join(repository, "apps/caudex-cli/README.md"), path.join(stage, "README.md"));
  if (args["sqlite-dll"]) await fs.copyFile(path.resolve(args["sqlite-dll"]), path.join(stage, "sqlite3.dll"));

  const metadata = {
    schemaVersion: 1,
    version: args.version,
    target: args.target,
    optimize: "ReleaseSafe",
    archiveStem: stem,
    platformCodeSigned: false,
    sqlite: {
      version: args["sqlite-version"],
      linkage: args["sqlite-linkage"],
      source: args["sqlite-source"],
      checksum: { algorithm: "sha256", value: args["sqlite-checksum"] ?? await sha256(path.resolve(sqliteFile)) },
    },
  };
  await fs.writeFile(path.join(stage, "build-metadata.json"), `${JSON.stringify(metadata, null, 2)}\n`);

  if (args.archive === "zip") await writeZip(stage, output, stem);
  else await writeTarGz(stage, output, stem);
  verifyArchive(await archiveEntries(output), stem, windows);
  smoke(stage, args.target, args.version);
} finally {
  await fs.rm(temporary, { recursive: true, force: true });
}

console.log(`packaged and smoke-tested ${output}`);

function parseArgs(values) {
  const parsed = {};
  for (let index = 0; index < values.length; index += 1) {
    const value = values[index];
    if (!value.startsWith("--")) throw new Error(`unexpected argument: ${value}`);
    const key = value.slice(2);
    if (index + 1 >= values.length || values[index + 1].startsWith("--")) throw new Error(`missing value for --${key}`);
    parsed[key] = values[++index];
  }
  return parsed;
}

function verifyArchive(entries, stem, windows) {
  const prefix = `${stem}/`;
  if (entries.some(({ name }) => !name.startsWith(prefix) || name.includes("\\") || name.split("/").includes(".."))) {
    throw new Error(`unsafe archive path in ${stem}`);
  }
  const files = new Set(entries.filter(({ name }) => !name.endsWith("/")).map(({ name }) => name.slice(prefix.length)));
  const required = new Set(["LICENSE", "NOTICE", "README.md", "build-metadata.json", windows ? "caudex.exe" : "caudex"]);
  if (windows) required.add("sqlite3.dll");
  if (files.size !== required.size || [...required].some((name) => !files.has(name))) {
    throw new Error(`unexpected files in ${stem}: ${[...files].sort().join(", ")}`);
  }
  const metadata = JSON.parse(readArchiveEntryFromEntries(entries, `${stem}/build-metadata.json`));
  if (metadata.archiveStem !== stem || metadata.platformCodeSigned !== false) throw new Error(`invalid archive metadata in ${stem}`);
}

function readArchiveEntryFromEntries(entries, wanted) {
  const entry = entries.find(({ name }) => name === wanted);
  if (!entry) throw new Error(`archive entry not found: ${wanted}`);
  return entry.data.toString("utf8");
}

function smoke(root, target, version) {
  const executable = path.join(root, target.endsWith("windows-gnu") ? "caudex.exe" : "caudex");
  if (target.endsWith("windows-gnu") && process.platform !== "win32") {
    chmodSync(executable, 0o755);
  }
  if (target.endsWith("windows-gnu") && !existsSync(path.join(root, "sqlite3.dll"))) throw new Error("Windows archive is missing sqlite3.dll");
  const environment = { ...process.env };
  if (target.endsWith("windows-gnu")) environment.PATH = root;
  const versionOutput = run(executable, ["version"], environment);
  if (!versionOutput.startsWith(`caudex ${version}\n`)) throw new Error(`unexpected CLI version output: ${versionOutput}`);
  const help = run(executable, ["--help"], environment);
  if (!help.includes("database backup") || !help.includes("history last")) throw new Error("extracted executable returned incomplete help");
  const database = path.join(root, "smoke.sqlite");
  const check = run(executable, ["--database", database, "--format", "json", "database", "check"], environment);
  if (!check.includes('"integrity":"ok"')) throw new Error(`SQLite smoke check failed: ${check}`);
  const started = run(executable, ["--database", database, "--format", "json", "workout", "start", "--command-id", "release-smoke-start", "--workout", "release-smoke-workout", "--started-at", "2026-01-01T00:00:00Z", "--occurred-at", "2026-01-01T00:00:00Z"], environment);
  if (!started.includes('"workoutId":"release-smoke-workout"')) throw new Error(`workout smoke start failed: ${started}`);
  const shown = run(executable, ["--database", database, "--format", "json", "workout", "show", "--workout", "release-smoke-workout"], environment);
  if (!shown.includes('"workoutId":"release-smoke-workout"')) throw new Error(`workout smoke read failed: ${shown}`);
}

function run(command, arguments_, environment) {
  const result = spawnSync(command, arguments_, { cwd: path.dirname(command), env: environment, encoding: "utf8" });
  if (result.status !== 0) throw new Error(`command failed (${result.status}): ${command} ${arguments_.join(" ")}\nstdout: ${result.stdout}\nstderr: ${result.stderr}`);
  return result.stdout;
}
